main/src/hadd.cxx

```diff
@@ -90,284 +102,473 @@
 #ifndef R__WIN32
 #include "ROOT/TProcessExecutor.hxx"
 #endif
 
 ////////////////////////////////////////////////////////////////////////////////
 
-int main( int argc, char **argv )
+inline std::ostream &Err() {
+   std::cerr << "Error in <hadd>: ";
+   return std::cerr;
+}
+
+inline std::ostream &Warn() {
+   std::cerr << "Warning in <hadd>: ";
+   return std::cerr;
+}
+
+inline std::ostream &Info() {
+   std::cerr << "Info in <hadd>: ";
+   return std::cerr;
+}
+
+using IntFlag_t = uint32_t;
+
+struct HAddArgs {
+   bool fNoTrees;
+   bool fAppend;
+   bool fForce;
+   bool fSkipErrors;
+   bool fReoptimize;
+   bool fDebug;
+   bool fKeepCompressionAsIs;
+   bool fUseFirstInputCompression;
+
+   std::optional<std::string> fWorkingDir;
+   std::optional<IntFlag_t> fNProcesses;
+   std::optional<TString> fCacheSize;
+   std::optional<ROOT::TIOFeatures> fFeatures;
+   std::optional<IntFlag_t> fMaxOpenedFiles;
+   std::optional<IntFlag_t> fVerbosity;
+   std::optional<IntFlag_t> fCompressionSettings;
+
+   int fOutputArgIdx;
+   int fFirstInputIdx;
+};
+
+enum class EFlagResult { kIgnored, kParsed, kError };
+
+static EFlagResult FlagToggle(const char *arg, const char *flagStr, bool &flagOut)
 {
-   if ( argc < 3 || "-h" == std::string(argv[1]) || "--help" == std::string(argv[1]) ) {
-         fprintf(stderr, kCommandLineOptionsHelp);
-         return (argc == 2 && ("-h" == std::string(argv[1]) || "--help" == std::string(argv[1]))) ? 0 : 1;
+   const auto argLen = strlen(arg);
+   const auto flagLen = strlen(flagStr);
+   if (argLen == flagLen && strncmp(arg, flagStr, flagLen) == 0) {
+      if (flagOut)
+         Warn() << "duplicate flag: " << flagStr << "\n";
+      flagOut = true;
    }
+   return flagOut ? EFlagResult::kParsed : EFlagResult::kIgnored;
+}
+
+// NOTE: not using std::stoi or similar because they have bad error checking.
+// std::stoi will happily parse "120notvalid" as 120.
+static std::optional<IntFlag_t> StrToUInt(const char *str)
+{
+   if (!str)
+      return {};
+
+   uint32_t res = 0;
+   do {
+      if (!isdigit(*str))
+         return {};
+      if (res * 10 < res) // overflow is an error
+         return {};
+      res *= 10;
+      res += *str - '0';
+   } while (*++str);
+
+   return res;
+}
+
+template <typename T>
+struct FlagConvResult {
+   T fValue;
+   EFlagResult fResult;
+};
+
+template <typename T>
+static FlagConvResult<T> ConvertArg(const char *);
+
+template <>
+FlagConvResult<std::string> ConvertArg<std::string>(const char *arg)
+{
+   return { arg, EFlagResult::kParsed };
+}
+
+template <>
+FlagConvResult<IntFlag_t> ConvertArg<IntFlag_t>(const char *arg)
+{
+   // Don't even try to parse arg if it doesn't look like a number.
+   if (!isdigit(*arg)) 
+      return { 0, EFlagResult::kIgnored };
+
+   auto intOpt = StrToUInt(arg);
+   if (intOpt)
+      return { *intOpt, EFlagResult::kParsed };
 
+   Err() << "error parsing integer argument '" << arg << "'\n";
+   return { {}, EFlagResult::kError };
```

Comment:
nitpick, but there is a `return { 0,` and a `return { {},` for both cases of returning an invalid int argument, maybe we should choose just one for consitency